src/navigation/partial_gate.py

```diff
@@ -0,0 +1,95 @@
+        """
+        # The Path will be (assuming the vector <x, y> from rover to post
+        # Point one: left perpendicular, <-y,x>/mag * Post_separation
+        # Point two: opposite end, <x,y>/mag * Post_separation
+        # Point three: right perpendicular, <y,-x>/mag * Post_separation
+        # Essentially traveling around the circle made by the first post
+
+        rover_to_post = post_pos - rover_pos
+        rover_to_post *= POST_SEPARATION / np.linalg.norm(rover_to_post)
+        # scale vector to have magnitude == POST_SEPARATION
+
+        left_perp = np.array([-rover_to_post[1], rover_to_post[0], 0])  # (-y,x)\
+        right_perp = np.array([rover_to_post[1], -rover_to_post[0], 0])  # (y,-x)
+
+        # This is just making our trajectory points into an array that we can read in
+        coords = np.vstack(
+            (post_pos + left_perp, post_pos + rover_to_post, post_pos + right_perp, post_pos - rover_to_post)
+        )
+        # should we add rover position to the end of this path? ^
+        # coords = np.hstack((coords, np.zeros(coords.shape[0]).reshape(-1, 1)))
+
+        return PartialGateTrajectory(coords)
+
+
+class PartialGateStateTransitions(Enum):
+    _settings_ = NoAlias
+    # State Transitions
+    no_fiducial = "SearchState"
+    partial_gate = "PartialGateState"
+    found_gate = "GateTraverseState"
+    done = "DoneState"
+
+
+class PartialGateState(BaseState):
+    def __init__(
+        self,
+        context: Context,
+    ):
+        super().__init__(context, add_outcomes=[transition.name for transition in PartialGateStateTransitions])  # type: ignore
+        self.traj: Optional[PartialGateTrajectory] = None
+
+    def evaluate(self, ud):
+        post_pos = self.context.env.current_fid_pos()
+        if post_pos is None:
+            post_pos = self.context.env.next_fid_pos()
+        gate = self.context.env.current_gate()
+
+        if gate is not None:  # If we have a gate, we are done
+            return PartialGateStateTransitions.found_gate.name  # type: ignore
+        elif post_pos is not None:  # Searching for second post
+            if self.traj is None:
+                self.traj = PartialGateTrajectory.partial_gate_traj(post_pos, self.context.rover.get_pose().position)
+        else:
+            return PartialGateStateTransitions.no_fiducial.name  # type: ignore
+
+        target_pos = self.traj.get_cur_pt()
+        cmd_vel, arrived = get_drive_command(
+            target_pos,
+            self.context.rover.get_pose(),
+            STOP_THRESH,
+            DRIVE_FWD_THRESH,
+        )
+        if arrived:
+            # if we finish the gate path, we're done (or continue search) CHECK THIS***
+            if self.traj.increment_point():
+                self.traj = None
+                self.context.course.increment_waypoint()
+                return PartialGateStateTransitions.done.name  # type: ignore
```

Comment:
just make sure you test that it does this. (maybe delete the other post entirely and see that it goes through the whole thing and then goes into done)

Reply:
works as expected :+1: